### PR TITLE
Add Arm Controller

### PR DESCRIPTION
**Pull request by Pokesi**

### CODE REVIEW

File: components/arm.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    ARM_END_MASS = 5  # kg
+    ARM_BASE_MASS = 1
+    ROTATE_GRAVITY_FEEDFORWARD = 1
+    EXTEND_GRAVITY_FEEDFORWARD = 1
+
+    MIN_EXTENSION = 0.7  # meters
+    MAX_EXTENSION = 1.3
+
+    ROTATE_GEAR_RATIO = (60 / 25) * (60 / 25) * (70 / 20)
+    EXTEND_GEAR_RATIO = (1 / 7) * (math.pi * 0.05)  # converts to meters
```
Comment (auscompgeek) [inline components/arm.py:21]: What is `0.05`? Could it have a name?

File: ids.py [hunk 2]
```diff
@@ -19,9 +19,18 @@ class Chassis:
...
+        extension = 15
 
-class PWM:
-    ...
+
+class PcmChannels:
+    arm_brake = 4
+
+
+class PwmPorts:
```
Comment (auscompgeek) [inline ids.py:32]: nit:
```suggestion
class PwmChannels:
```

File: components/arm.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        self.extension_encoder.setPositionConversionFactor(self.EXTEND_GEAR_RATIO)
+        self.extension_encoder.setVelocityConversionFactor(self.EXTEND_GEAR_RATIO * 60)
+        # assume retracted starting position
+        self.extension_encoder.setPosition(self.MIN_EXTENSION)
+        self.extension_controller = ProfiledPIDController(
+            1, 0, 0, TrapezoidProfile.Constraints(maxVelocity=2, maxAcceleration=5)
+        )
+        self.extension_simple_ff = SimpleMotorFeedforwardMeters(0.1, 0.5, 1)
+
+        self.absolute_encoder = DutyCycleEncoder(PwmPorts.arm_abs_encoder)
```
Comment (auscompgeek) [inline components/arm.py:72]: Out of curiosity, is there a particular reason we're wiring this to the roboRIO rather than the motor controller?

**Review by auscompgeek — CHANGES_REQUESTED** [review]

File: components/arm.py (hunk 1 above)
Comment (OliverW10) [reply components/arm.py:21]:
```suggestion
    SPOOL_DIAMETER = 0.05
    EXTEND_GEAR_RATIO = (1 / 7) * (math.pi * SPOOL_DIAMETER)  # converts to meters
```

**Review by auscompgeek — COMMENTED** [review]
I don't think there's anything obviously problematic here, but I do have a couple of comments.

File: components/arm.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        self.rotation_motor = rev.CANSparkMax(
+            CanIds.Arm.rotation_left, rev.CANSparkMax.MotorType.kBrushless
+        )
+        self.rotation_motor.setIdleMode(rev.CANSparkMax.IdleMode.kBrake)
+        self.rotation_motor.setInverted(False)
+        # setup second motor to follow first
+        self._rotation_motor_right = rev.CANSparkMax(
+            CanIds.Arm.rotation_right, rev.CANSparkMax.MotorType.kBrushless
+        )
+        self._rotation_motor_right.follow(self.rotation_motor, True)
```
[the remark this answers is not in the document]
Comment (OliverW10) [reply components/arm.py:48]: [to invert](https://robotpy.readthedocs.io/projects/rev/en/stable/rev/CANSparkMax.html#rev.CANSparkMax.follow)

File: components/arm.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    ARM_BASE_MASS = 1
+    ROTATE_GRAVITY_FEEDFORWARD = 1
+    EXTEND_GRAVITY_FEEDFORWARD = 1
+
+    MIN_EXTENSION = 0.7  # meters
+    MAX_EXTENSION = 1.3
+
+    ROTATE_GEAR_RATIO = (60 / 25) * (60 / 25) * (70 / 20)
+    SPOOL_DIAMETER = 0.05
+    EXTEND_GEAR_RATIO = (1 / 7) * (math.pi * SPOOL_DIAMETER)  # converts to meters
```
[the remark this answers is not in the document]
Comment (OliverW10) [reply components/arm.py]: Changed to 'spool output ratio'

File: components/arm.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+        # TODO: get pid and feedforward values for arm and extension from sysid
+        # running the controller on the rio rather than on the motor controller
+        # to allow access to the velocity setpoint for feedforward
+        rotation_constraints = TrapezoidProfile.Constraints(
+            maxVelocity=3, maxAcceleration=3
+        )
+        self.rotation_controller = ProfiledPIDController(5, 0, 0, rotation_constraints)
+        # ignore 'meters', theres no generic one
+        self.rotation_simple_ff = SimpleMotorFeedforwardMeters(0.1, 0.5, 1)
```
Comment (auscompgeek) [inline components/arm.py:70]: Just curious: why not ArmFeedforward?

**Review by auscompgeek — APPROVED** [review]

**Pull request merged**